cdap-app-templates/cdap-etl/cdap-data-streams-base/src/test/java/io/cdap/cdap/datastreams/DataStreamsTest.java

```diff
@@ -1436,4 +1437,106 @@ private long getMetric(ApplicationId appId, String metric) {
+
+    // since dataset name is a macro, the dataset isn't created until it is needed. Wait for it to exist
+    Tasks.waitFor(true, () -> getDataset(outputName).get() != null, 3, TimeUnit.MINUTES);
+
+    DataSetManager<Table> outputManager = getDataset(outputName);
+    Tasks.waitFor(
+      true,
+      () -> {
+        outputManager.flush();
+        Set<StructuredRecord> outputRecords = new HashSet<>(MockSink.readOutput(outputManager));
+        return expected.equals(outputRecords);
+      },
+      3,
+      TimeUnit.MINUTES);
+
+    MockSink.clear(outputManager);
+    sparkManager.stop();
+    sparkManager.waitForStopped(10, TimeUnit.SECONDS);
+    RecoveringTransform.reset();
+  }
+
+  @Test
+  /**
+   * Uses {@link RecoveringTransform} plugin and custom retry settings to test retry timeout.
+   */
+  public void testRetryTimeout() throws Exception {
+    Schema schema = Schema.recordOf(
+      "retry_timeout_test",
+      Schema.Field.of("id", Schema.of(Schema.Type.STRING)),
+      Schema.Field.of("name", Schema.of(Schema.Type.STRING))
+    );
+    List<StructuredRecord> input = new ArrayList<>();
+    StructuredRecord samuelRecord = StructuredRecord.builder(schema).set("id", "123").set("name", "samuel").build();
+    StructuredRecord jacksonRecord = StructuredRecord.builder(schema).set("id", "456").set("name", "jackson").build();
+    input.add(samuelRecord);
+    input.add(jacksonRecord);
+
+    DataStreamsConfig etlConfig = DataStreamsConfig.builder()
+      .addStage(new ETLStage("retry_timeout_source", MockSource.getPlugin(schema, input)))
+      .addStage(new ETLStage("retry_timeout_sink", MockSink.getPlugin("${retry_timeout_output_macro}")))
+      .addStage(new ETLStage("retry_timeout_transform", RecoveringTransform.getPlugin()))
+      .addConnection("retry_timeout_source", "retry_timeout_transform")
+      .addConnection("retry_timeout_transform", "retry_timeout_sink")
+      .setBatchInterval("1s")
+      .build();
+
+    ApplicationId appId = NamespaceId.DEFAULT.app("simpleRetryTimeoutApp");
+    AppRequest<DataStreamsConfig> appRequest = new AppRequest<>(APP_ARTIFACT, etlConfig);
+    ApplicationManager appManager = deployApplication(appId, appRequest);
+
+    Set<StructuredRecord> expected = new HashSet<>();
+    expected.add(samuelRecord);
+    expected.add(jacksonRecord);
+
+    SparkManager sparkManager = appManager.getSparkManager(DataStreamsSparkLauncher.NAME);
+    // Timeout retry in 1 min and set the base delay to 60s.
+    sparkManager.start(ImmutableMap.of("retry_timeout_output_macro", "retry_timeout_output",
+                                       "cdap.streaming.maxRetryTimeInMins", "1",
+                                       "cdap.streaming.baseRetryDelayInSeconds", "60"));
+    sparkManager.waitForRun(ProgramRunStatus.RUNNING, 10, TimeUnit.SECONDS);
+    // Should fail after retry times out
+    sparkManager.waitForRun(ProgramRunStatus.FAILED, 2, TimeUnit.MINUTES);
```

Comment:
it would be good to also check that a state handling source does not write any state in this situation.

Reply:
Will add this in the plugin repo as discussed offline.